src/OAuth/Repository/User/MongoUserRepository.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+namespace Cerberus\OAuth\Repository\User;
+
+use Cerberus\Collection\PaginatedCollection;
+use Cerberus\Exception\EntityNotFoundException;
+use Cerberus\Hasher\HasherInterface;
+use Cerberus\OAuth\User;
+use Doctrine\Common\Persistence\ObjectRepository;
+use Doctrine\ODM\MongoDB\DocumentManager;
+use League\OAuth2\Server\Entities\ClientEntityInterface;
+use League\OAuth2\Server\Entities\UserEntityInterface;
+use Pagerfanta\Adapter\DoctrineODMMongoDBAdapter;
+use Pagerfanta\Pagerfanta;
+
+class MongoUserRepository implements UserRepositoryInterface
+{
+    /**
+     * @var DocumentManager
+     */
+    private $manager;
+    /**
+     * @var HasherInterface
+     */
+    private $hasher;
+    /**
+     * @var ObjectRepository
+     */
+    private $repository;
+    /**
+     * @var Pagerfanta
+     */
+    private $paginator;
+
+    public function __construct(DocumentManager $manager, HasherInterface $hasher)
+    {
+        $this->manager = $manager;
+        $this->hasher = $hasher;
+        $this->repository = $this->manager->getRepository('Cerberus:User');
+        $queryBuilder = $this->manager->createQueryBuilder('Cerberus:User');
+        $mongoAdapter = new DoctrineODMMongoDBAdapter($queryBuilder);
+        $this->paginator = new Pagerfanta($mongoAdapter);
+    }
+
+    /**
+     * Get a user entity.
+     *
+     * @param string $username
+     * @param string $password
+     * @param string $grantType The grant type used
+     * @param ClientEntityInterface $clientEntity
+     *
+     * @return UserEntityInterface|false
+     */
+    public function getUserEntityByUserCredentials(
+        $username,
+        $password,
+        $grantType,
+        ClientEntityInterface $clientEntity
+    ) {
+        /** @var User $user */
+        $user = $this->repository->findOneBy(['username' => $username]);
+
+        if (! $user) {
+            return false;
+        }
+
+        if (! $this->hasher->verify($password, $user->getPassword())) {
+            return false;
+        }
+
+        return $user;
+    }
+
+    /**
+     * @param string $id
+     * @return User
+     * @throws EntityNotFoundException
+     */
+    public function find(string $id): User
+    {
+        /** @var User $user */
+        $user = $this->repository->find($id);
+
+        if (! $user) {
+            throw EntityNotFoundException::create(User::class, $id);
+        }
+
+        return $user;
+    }
+
+    /**
+     * @param User $user
+     */
+    public function save(User $user, User ...$users): void
```

Comment:
This function will not fail without enforcing the first user. This would increase the readability, especially since the `$user` variable is being re-used in the loop.

Reply:
Agreed. Will fix!